src/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -514,11 +502,18 @@ internal static char ParseFormatSpecifier(string format, out int digits)
             return (char)0; // Custom format
         }
 
-        private static string FormatBigIntegerToHexString(BigInteger value, char format, int digits, NumberFormatInfo info)
+        private static string FormatBigIntegerToHex(BigInteger value, char format, int digits, NumberFormatInfo info)
         {
-            StringBuilder sb = new StringBuilder();
-            byte[] bits = value.ToByteArray();
-            string fmt = null;
+            Debug.Assert(format == 'x' || format == 'X');
+
+            // Get the bytes that make up the BigInteger.
+            Span<byte> bits = stackalloc byte[64]; // arbitrary limit to switch from stack to heap
+            bits = value.TryWriteBytes(bits, out int bytesWritten) ?
+                bits.Slice(0, bytesWritten) :
+                value.ToByteArray();
+
+            Span<char> stackSpace = stackalloc char[128];
+            var sb = new ValueStringBuilder(stackSpace);
```

Comment:
given the semantics are not quite the same, and ToString doesn't work, maybe `vsb` instead of `sb`?